cs169/BJC-Teacher-Tracker-App
# Fix email template reseed problem upon updates/deletion

### [Pivotal Tracker Link][tracker]


[tracker]: https://www.pivotaltracker.com/story/show/187113121

## What this PR does:


This pull request ___fixes___ the issue of duplication of Email Templates upon reseeding the database with updated seed data. Based on the special needs from Professor Ball, it does not purge old email templates that are not present in the current seed data.

### Include screenshots, videos, etc.

#### Special needs from Professor Ball
<img width="1095" alt="image" src="https://github.com/cs169/BJC-Teacher-Tracker-App/assets/92573736/908832a5-930e-4fd9-826c-f178431a4560">

#### Previous Problem - Duplication of Email Templates after Reseeding Updated Email Templates
![Previous Problem](https://github.com/beautyjoy/BJC-Teacher-Tracker/assets/92573736/3091bd23-73fa-407a-a5d2-a98b2a8f9b2d)

#### After Fix
1. No Duplication of Email Templates after Reseeding Updated Email Templates.
2. It also considers the case to not purge old email templates that are not present in the current seed data.

#### Who authored this PR?

1. Perry (Jingchao) Zhong
2. Michael Tao

### How should this PR be tested?

- Ensure that the test suite covers scenarios for both updating existing email templates and not-removing old templates from the database upon reseeding.
- Specifically, the tests should verify that after reseeding:
  1. Updated templates reflect the new changes without creating duplicate entries.
  2. Templates in the db, but not included in the new seed data, should not be removed.

* **What do the specs/features test?**
  The specs test the reseeding process's ability to update email template (with same hash) and keep old unchanged email templates in the database accurately.

* **Are there edge cases to watch out for?**
  Not on my mind yet.

#### Are there any complications to deploying this?
This PR is basically safe. If there are any potential risks, the update feature might overwrite the old email templates in the database if the hashed values are the same.

[Relevant code of hashing can be found here](https://github.com/cs169/BJC-Teacher-Tracker-App/pull/34/files#diff-c61a903a1201b0c7f076b0270377c20390c26f655ce63689d7d843296fcf63f3R5-R9)

### Checklist:

- [x] I have renamed the branch to match PivotTracker's suggested one (necessary for BlueJay) (e.g., `perry/187113121-fix-email-template-reseed-issue`). Any branch name will do as long as the story ID is there. You can use `git checkout -b [new-branch-name]`
